Add screenshot

qxprakash opened the pull request:
# Screenshot Support for Screen Capture Plugin

Closes #3310

## Overview

Added screenshot capture functionality to `@uppy/screen-capture` plugin, allowing users to take screenshots alongside screen recordings.

### API Changes
Added new configuration options:
- `enableScreenshots`: Toggle screenshot functionality (default: false)
- `screenshotQuality`: Control image compression (0.0-1.0, default: 0.9)
- `preferredImageMimeType`: Set output format ('image/jpeg'|'image/png', default: 'image/jpeg')

### Implementation Details
- Reused existing screen capture stream for drawing screenshot and converting them to Blob and storing the image
- proper state cleanup , in case of error and component unmount
- added missing component prop types and return types in RecordButton and SubmitButton Components 

Hi @Murderlon let me know what do you think about this, also how do I update the docs for this plugin ? and the stackblitz live example.

## Screenshots
![image](https://github.com/user-attachments/assets/c68c4746-fdf8-447d-87c3-ff6c1b517637)

qxprakash commented:
@Murderlon , here are few of my questions relating to the PR 

1) In order to handle validations for params take for example `preferredImageMimeType` when the user provides an invalid value (let's suppose a random string instead of a correct mime type) , I was confused between , throwing an error and letting user know that they've entered a wrong value **vs** just using a fallback value , 

Both approaches have their own sets of pros and cons:

  - In case of an error , the onus is on the user to read the docs carefully and use correct values.
  - In case of a fallback, the user might get a false sense of things working as they want, but in reality, it's not as we're using a fallback value.

I went with the second approach.

2) I couldn't find any existing unit tests for this plugin ?
3) Do you think it makes sense for adding a preview functionality similar to the screencast video ?

Murderlon reviewed (commented):
Overall looks good 👌 

Some comments:
- I think we should replace both icons in the button with `camera` and `video-camera` from https://heroicons.com/solid
- I'm not sure if the `screenshotQuality` option makes sense to add, can't we always do max quality? If people want compression you can add `@uppy/compressor`. Bit of a conflicting feature with that plugin.
- Should we make `enableScreenshots` default to `true`? I'm open to both.
- I think fallback makes sense when an incorrect mime type is given to align with how video does it.

> I couldn't find any existing unit tests for this plugin ?

Quite a few plugins aren't tested so that's fine. I don't think we need it for this feature.

> Do you think it makes sense for adding a preview functionality similar to the screencast video ?

That would be nice, yes!

qxprakash commented:
### A couple more things I thought about:

- Once the user has taken a screenshot and it’s in preview, there’s no point in showing the screenshot button again — there's really no reason to take an identical screenshot. So, I removed it.

- If the user has started recording, the screenshot button should be hidden. Keeping it visible could lead to some edge cases, such as:
  - The user taking a screenshot while the recording is still live.
  - User taking a screenshot when video is in preview mode

So, I’ve disabled the screenshot button in the above cases.

---

### TL;DR

- When recording starts, the screenshot button is hidden until the user either uploads or discards the recording.
- Once a screenshot has been taken and is in preview, the screenshot button also disappears.

Murderlon reviewed (commented):
Code LGTM

> * Once the user has taken a screenshot and it’s in preview, there’s no point in showing the screenshot button again — there's really no reason to take an identical screenshot. So, I removed it.

I think it's currently a bit confusing, after a screenshot the button that is auto focussed is to start a recording. This seems counter intuitive after just taking a screenshot. I'm wondering if we should shared submit/discard button combination after taking a screenshot or taking a video. Submit closes the plugin, discard returns to the button combo record/screenshot.

What do you think?

Murderlon commented on examples/svelte-example/package.json:
```diff
@@ -23,13 +23,13 @@
-    "@uppy/svelte": "workspace:*",
-    "@uppy/webcam": "workspace:*",
-    "@uppy/xhr-upload": "workspace:*",
+    "@uppy/core": "^4.0.0",
```
Unrelated change, let's revert this

qxprakash replied:
This was probably me trying to modify the examples to get the StackBlitz embed working on the uppy.io repo, but it ended up in this PR unexpectedly lol , I've reverted it now.

qxprakash commented:
Thanks for the review !
> 
> > * Once the user has taken a screenshot and it’s in preview, there’s no point in showing the screenshot button again — there's really no reason to take an identical screenshot. So, I removed it.
> 
> I think it's currently a bit confusing, after a screenshot the button that is auto focussed is to start a recording. This seems counter intuitive after just taking a screenshot. I'm wondering if we should shared submit/discard button combination after taking a screenshot or taking a video. Submit closes the plugin, discard returns to the button combo record/screenshot.
> 
> What do you think?

Yup, agreed!
I’ve added a cancel button to discard the media and return to the previous state, where the user is presented with both the screenshot and record buttons. This is much more consistent.”

![image](https://github.com/user-attachments/assets/23d49a7c-b3bf-4339-81f6-6b944f805f36)

Murderlon reviewed (approved):
All good now!

One last suggestion: swap the order of discard and submit. My gut feeling is that having the autofocus and visual order on submit is more intuitive.

Then ready to merge
